tinyusb: Support USB device drivers written in Python

projectgus opened the pull request:
The goal is to allow implementing USB device drivers in Python via TinyUSB at runtime, in addition to any existing compile-time TinyUSB class drivers (i.e. CDC and optionally MSC).

Python API with the more interesting user-facing code is at https://github.com/micropython/micropython-lib/pull/558 - if checking out this branch in micropython, go to the `lib/micropython-lib` subdirectory and check out the `feature/usbd_python` branch there as well. 

## Port Support

This PR implements this for the `rp2` and `samd` ports.

The port-level requirement is that the TinyUSB task is run in a "scheduled" static callback node, as implemented in #12846 for rp2 & samd.) I think this approach can be extended to most of the other ports pretty easily.

Challenging ports (future work):

* esp32. ESP-IDF runs TinyUSB in its own FreeRTOS Task. I think it can be made to work here by using a cross-task Python callback (see jimmo's [implementation here](https://github.com/micropython/micropython/pull/8945/commits/843204c717b7c8446398fffe4ebcdb0bd13765df)), or by changing the TinyUSB implementation to work similar to the other ports (this will save some RAM and on single core ESP32s it might even improve performance. On dual core it might reduce performance a little but it also might not.)
* stm32. Currently doesn't use TinyUSB. It may be possible to port this approach to the ST USB stack, but prefer to make a compile-time option to build stm32 port with TinyUSB. This will be less work and more maintainable into the future.
* zephyr. The Zephyr [USB device driver API](https://docs.zephyrproject.org/latest/connectivity/usb/device/api/usb_device.html) (example: [hid core](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/device/class/hid/core.c)) looks like it would adapt to this approach easily, but haven't verified this yet.

## Binary Size Impact

Where possible, the implementation has been placed into micropython-lib to reduce the baseline binary size impact. If enabled at compile time (currently not the default) the code size change is:

| Port | Board | Code Size  | Delta | Static RAM Size| Delta |
| ------ | -------- | -------------------- | ------- | ------------ |  ---- |
| rp2 | RPI_PICO | 322808 -> 325704 | +2896| 5872 -> 5876 | +4 |
| samd | SEEED_WIO_TERMINAL (samd51) | 261136 -> 263984 | +2848 | 3068 -> 3072 | +4 |

Note: This doesn't include freezing in the "usbd" module from micropython-lib. Currently that module is approximately 15KB.

## How To Use

* Edit the `mpconfigport.h` file for your port or board and change `MICROPY_HW_ENABLE_USB_RUNTIME_DEVICE` to 1.
* (Recommended) edit the`manifest.py` file for your port or board and insert `require("usbd")`.
* Rebuild the firmware

See https://github.com/micropython/micropython-lib/pull/558 for more.

## Debugging Tips

REPL debugging via the USB serial port is obviously perilous when the USB stack itself is being debugged, and the whole device needs to re-enumerate regularly for the host to see changes.

### Hardware UART

Switch the REPL interface to a hardware UART if you can (for rp2, set `MICROPY_HW_ENABLE_UART_REPL` in `ports/rps2/mpconfigport.h`.) This makes `mpremote mount` very slow, but it means crashes that take down USB are still reported on the REPL!

### TinyUSB debug

To enable TinyUSB debug output, add the lines `#undef CFG_TUSB_DEBUG` and `#define CFG_TUSB_DEBUG 2` (or a different level) to the `shared/tinyusb/tusb_config.h` header. Debugging output goes to stdout, so don't do this unless you've also switched to a hardware UART.

### Semihosting

Another option is semihosting. On rp2 with an SWD debugger connected, can [apply this quick patch](https://gist.github.com/projectgus/55a5a155c87b8fd424fe580d130d1008) and then run pyocd as follows:

```
pyocd gdbserver --semihosting --target rp2040 --frequency 32m --persist 
```

Read from localhost port 4444 and you'll see stdout at any time that USB-CDC device is not open. (This failover behaviour is useful because semihosting is painfully slow, and also pyocd doesn't support non-UTF-8 input or output so any "raw REPL" traffic will cause it to crash.)

In recent testing the semihosting patch is still not stable when using `mpremote` to run Python code that triggers re-enumeration, it panics trying to write to the CDC while the device is being reset (more debugging needed).

*This work was funded through GitHub Sponsors.*

projectgus commented:
Have updated this branch a bit, and also submitted a draft set of micropython-lib changes at https://github.com/micropython/micropython-lib/pull/558

turmoni commented:
I've decided I like pain and so I'm writing a barebones read-only mass storage driver for my use case using this work (not pain because the work's hard to use, just because now I have to understand USB and SCSI at the same time). I'm trying to figure out if it's possible to stall an endpoint with the current implementation? I believe that's what I'm meant to do in the case that I get an invalid request, but I'm not seeing that that's functionality that's exposed. However, I also don't really understand what I'm talking about, so stalling an endpoint might actually be a synonym for something else that is exposed.

turmoni commented:
I now understand slightly more, I think, so I can give a better overview. Mass storage mostly works via data being sent through the IN and OUT endpoints, with a couple of exceptions (literally two, and one of those is reset). The correct response to an invalid request coming in on OUT is to stall IN, but I'm not seeing how I'd be able to do this. Now, I have got what I need working on Linux without doing this, but I don't know how robust it is.

projectgus commented:
@turmoni Great plan!

I think you are correct, there are 3 TinyUSB usbd_priv.h interface function calls: `usbd_edpt_stall`, `usbd_edpt_clear_stall`, and `usbd_edpt_stalled` that are not exposed via the Python API yet, but will need to be in order to implement the Mass Storage class driver and other cases.

I haven't looked closely as MSC yet. My vague plan was to work up in complexity from HID (mostly done), to MIDI (unexpectedly complex), to CDC, and then MSC (complex because SCSI). However, from a very quick look at the [TinyUSB class driver](https://github.com/hathach/tinyusb/blob/868f2bcda092b0b8d5f7ac55ffaef2c81316d35e/src/class/msc/msc_device.c) it seems like stalling the endpoint is needed for two cases:

1. The host sends an invalid request via USB.
2. The SCSI storage "thing" is not ready yet or is otherwise in a bad state.

So if your particular implementation is happy to ignore (1) or handle it in a more extreme way, and if it doesn't need (2) because the read-only backing storage is always available, you might be alright without endpoints stalling.

If you do want to add those APIs to the Python interface and feel like sending a PR to add them to this branch, then that would also be great - they will certainly be needed before this work can be completed.

turmoni commented:
I think I _should_ be OK without having to implement them for my needs. I've got my device working as much as I need it to on Linux, just Windows is causing me issues. Specifically, I keep getting `maximum recursion depth exceeded` and I'm trying to figure out where that's coming from so I can avoid it. I _suspect_ it's some kind of issue where the callbacks are happening so quickly that the function that initiated the transfer hasn't finished yet, but I've yet to properly track it down.

Before realising I had some UARTs I could hook up for better logging I did get what I thought was weird corruption on my two-line LCD when I hit the recursion depth error, but looking at more verbose output over my UART I think it's just supporting my idea that it's the callback happening quickly.

I don't know if this may have been something you ran into at any point? I _can_ provide my code that currently reliably has this issue for me on Windows 10, but it's currently between the states of "nice tidy code that has been written before testing any of it" and "reasonable somewhat tidy code that has been debugged and cleaned up", so it's now at "horrible hacky mess with various attempts at making things work shoehorned in"

turmoni commented:
Thinking about it, I might as well just attach it here regardless - I do intend to add it to turmoni/micropython-lib when it's in a better (working, cleaner, more documented) state, but with the caveats that I know it's broken and in the middle of debugging I'm not too embarrassed to share the code.

To try it out, create an `MSCInterface` object as you have for the other examples, `filesystem` is either None or a `bytes`-like object (I'm currently feeding it a tiny FAT filesystem). `uart` and `lcd` are optional parameters for debugging (I suspect the logging might be contributing to the recursion depth error). I have the USB interface and the SCSI interface abstracted away from one another in a way that the TinyUSB code doesn't, but hopefully this makes it a bit easier to follow.

The basic flow is:

- Standard setup stuff happens, host asks for the number of LUNs on the control channel.
- The device prepares to accept a CBW (which contains the command) on the OUT endpoint
- On receiving a CBW, the device validates it, sends any data that's been requested on the IN endpoint, then submits a CSW (status) report on the IN endpoint
- When the CSW has been sent, we start again from the preparation to accept the CBW

What I'm seeing on Windows is that it's mostly working fine, until it ends up failing to do something, and then I don't think I have the ability to get it back in sync, but ideally I shouldn't need to.

(I don't think anything ever hits handle_endpoint_control_xfer, so feel free to ignore that.)
[brokenmsc.txt](https://github.com/micropython/micropython/files/11806403/brokenmsc.txt)

projectgus commented:
Thanks @turmoni. At a glance this looks quite good to me! I haven't had time to run it, might be a while before I have that time.

I don't think callback frequency by itself would trigger a recursion depth issue: the callbacks should all be called from the `tud_task_ext` "main loop task function" of TinyUSB, which is itself called from the usbd_task() method in shared/tinyusb/mp_usbd.c, which is called from `MICROPY_EVENT_POLL_HOOK_FAST`. That code should never be re-entrant, it handles one DCD (low-level peripheral) event at a time. Therefore, even if one transfer callback did something that triggered an immediate new callback at the hardware level then you'd still expect the inner callback to fully return all the way back through `tud_task_ext`, before MicroPython keeps running and hits the next  event hook causing `tud_task_ext` to be called again and the new event to trigger from the top.

The default stack size for rp2 port is [8KB](https://github.com/micropython/micropython/blob/master/ports/rp2/CMakeLists.txt#L434), set via `PICO_STACK_SIZE`. Is it possible you're actually exceeding this, without infinite recursion? The maximum depth will be determined by the total stack depth of the "main" Python code, plus the potential "event hook" callback stack frames on top of that from a single callback event. 8KB sounds like heaps, but actually it might be getting up there...? If you're able to switch the REPL to the UART, you should get a decent stack trace when this triggers (I hope!)

projectgus commented:
OK, what I confidently wrote just before is totally wrong - there are currently a bunch of places where an event hook may call back into `tud_task_ext()`, which is quite a nasty bug with some nasty implications!

For most cases we can put an anti-recursion check into `tud_task_ext()`, but there are some places even this won't work (like if a callback tries to write to a USB device like the USB-CDC and that relies on event processing to make space available.)

So, it may have to be one of: treat USB xfer callbacks almost like hard interrupts and do no significant work in them, *or* change the callbacks to be called via `mp_sched_schedule` instead of `mp_call_function_n_kw` meaning they get "re-queued" and handled there. This is a little worse for performance, but it might end up being the way it has to be.

turmoni commented:
Thanks, I did quickly do a new firmware build with the REPL on UART 0 and that didn't seem to enlighten me any further, so I suspect it is some kind of unexpected interaction. I had a quick look to see if I could hackily change how the callbacks are called, but I'm neither proficient enough at C nor the micropython codebase to be able to do anything useful there in the short time I have to look at it this evening.

turmoni commented:
Right, so, I've done terrible things to your code and I feel dirty, but I think I understand one of the issues I'm having. Firstly, I replaced all my callbacks with short functions that called micropython.schedule() to schedule something that actually does the work, so hopefully if that's an issue, that's no longer a factor with what I'm doing. This still didn't fix it though, so I did some more digging.

I've copied and pasted various helper functions from the micropython source to try to figure out what the issue is that I'm now facing, and I think that it's `_submit_xfer` is too slow to add the callback (or something along those lines). My horrific debug output is as follows:

```
About to submit xfer
Submitting 55534253020000000000000000, length: 13
Result 1
In runtime_dev_xfer_cb, ep_addr: 133, result: 0, xferred: 13
Calling callback
object <bound_method> is of type bound_method
<bound_method><bound_method 2001c5f0 <_USBDevice object at 2001a5e0>.<function _xfer_cb at 0x20013380>>
In _xfer_cb!
itf: <MSCInterface object at 2001bc00>, cb: None
It's not none
Submitting xfer - 133, 55534253020000000000000000. itf = <MSCInterface object at 2001bc00>, cb = <bound_method>
Submitted xfer?
```
So one of the first things I realised was that I felt silly because I didn't actually have to do anything with the C code, because all it did was point me to the fact that I should be looking at _xfer_cb because that's the thing that actually invokes the right Python callback. But this does (probably) show the overall flow as it happened; line-by-line:

- `_submit_xfer` is called (Python)
- This calls `_usbd.submit_xfer`, and the C code is printing the value it's sending and the length
- "Result 1" is the result from calling `usbd_edpt_xfer`
- "In runtime_..." is the callback handler in the C side
- "Calling callback" comes after the `mp_obj_is_callable` check, and is followed by a couple of red herrings I didn't need to care about
- "In _xfer_cb!" is back in the Python code
- "itf: ..., cb: None" *is the important part* - at the time that the callback is invoked, it doesn't think there's a user-defined callback to call
- "It's not none" is back in the C code and can be ignored
- "Submitting xfer" is back in the Python code, in `_submit_xfer`, and is a print statement I put in that happens just before the interface and callback are set.
- "Submitted xfer?" is just before `_submit_xfer` returns True

This makes a lot of sense for the non-recursion-error symptoms I'm seeing, if the callback handler is being called before the callback has had a chance to be written.

turmoni commented:
Well having worked around this (see below) it looks like it was just an error I was exposing for myself, and it wasn't the original problem I was having, where my CSW claims to have been submitted with the right data but goes missing and doesn't make it to my USB capture in Wireshark. I guess that's tomorrow's problem for me to investigate.

```python
    def _retry_xfer_cb(self, args):
        print("Retry!")
        (ep_addr, result, xferred_bytes) = args
        self._xfer_cb(ep_addr, result, xferred_bytes)

    def _xfer_cb(self, ep_addr, result, xferred_bytes):
        # Singleton callback from TinyUSB custom class driver when a transfer completes.
        print("In _xfer_cb!")
        try:
            itf, cb = self._eps[ep_addr]
            print(f"itf: {itf}, cb: {str(cb)}")
            if cb is None:
                micropython.schedule(self._retry_xfer_cb, (ep_addr, result, xferred_bytes))
                return

            self._eps[ep_addr] = (itf, None)
        except KeyError:
            cb = None
        if cb:
            print("Got a cb")
            cb(ep_addr, result, xferred_bytes)
```

turmoni commented:
Right, I think I've solved my magical disappearing CSW problem, and that aspect was my fault; as per the spec:

- If the device intends to send less data than the host indicated, then:
- - The device shall send the intended data.
- - - The device may send fill data to pad up to a total of dCBWDataTransferLength.
- - - If the device actually transfers less data than the host indicated, then:
- - - - The device may end the transfer with a short packet.
- - - - The device shall STALL the Bulk-In pipe.
- - The device shall set bCSWStatus to 00h or 01h.
- - The device shall set dCSWDataResidue to the difference between dCBWDataTransferLength and the actual amount of relevant data sent.

I was sending less than dCBWDataTransferLength, but not STALLing (since we can't do that yet). Padding out the response to dCBWDataTransferLength seems to solve my issue. Now to go and see how much of this code I can rip out again...

turmoni commented:
I'm now trying to set up a composite device, using both HID and MSC, and my brain has been slightly fried. It looks like whichever interface I `add_interface` first is the only one that gets control callbacks fired for it - I can see in my USB capture that my host is sending a control transfer with a wIndex of 3, but that doesn't seem to make it through to `runtime_dev_control_xfer_cb` in the C code. Since there's only one relevant request that comes through for mass storage, I can somewhat fudge it and make both work for a bit for me by just initialising the endpoint regardless, but I don't know how robust this is.

(Another issue I ran into is with the Python side where the current logic for picking endpoints doesn't work because once you have an IN endpoint >=0x80, adding one to it won't get you back into OUT endpoint territory, but I'm putting that here more as a note for anyone else trying this.)

turmoni commented:
OK so I've now done everything I needed to; to summarise the things I'm working around:
 * The inability to STALL endpoints
 * Control transfer callbacks not firing for anything other than the first interface that's added
 * Potential race condition where a callback might fire before the specific callback handler has been registered in `self._eps`
 * Potential issues with callbacks being weird due to being directly called rather than scheduled - I *think* the root of my actual issue there may have been with me not sending the right data, mind
 * Not really seeing a way to handle input on the control endpoint
 
To elaborate on that last one, since I hadn't mentioned it earlier, USB HID allows you to either receive OUT-style requests on the OUT endpoint if you've defined one, or on the control endpoint. I spent a little bit of time trying to figure out how you'd make this work generically before deciding it was easier to just add an OUT endpoint to USB HID for my purposes.

Thanks for doing this work, I'm glad I didn't have to do all my logic in C!

projectgus commented:
@turmoni Thanks for working through this in so much detail. I'm going to look into some of these items soon and get back to you.

projectgus commented:
This PR has been rebased on master, but without any other significant code changes yet.

projectgus commented:
Rebased again to remove the commit that updates the micropython-lib submodule. I'll try to avoid rebasing from here on to stop breaking history for others, can do a final squash once it's closer to being finished. If checking this out then you'll need to manually check out the matching branch in micropython-lib.

@turmoni I've gone through the list of issues mentioned above (thanks again for documenting these!) and made some changes, will comment shortly on the micropython-lib PR with details as that's where most of the changes ended up being needed.

projectgus commented:
Rebased, made a bunch of fixes around transfer cleanup and how exceptions in USB device callbacks are handled. (These are an inevitability on ports like rp2 which call into TinyUSB from a hook, because `KeyboardError` can be injected into a Python callback.)

projectgus commented:
Rebased and pushed some more code changes, including some changes to the C<->Python interface. Will explain more on the linked micropython-lib PR.

projectgus committed:
py/objstr: Add a macro to define a bytes object at compile time.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
extmod/machine_usb_device: Add support for Python USB devices.

This new machine-module driver provides a "USBDevice" singleton object and
a shim TinyUSB "runtime" driver that delegates the descriptors and all of
the TinyUSB callbacks to Python functions.  This allows writing arbitrary
USB devices in pure Python.  It's also possible to have a base built-in
USB device implemented in C (eg CDC, or CDC+MSC) and a Python USB device
added on top of that.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
samd: Enable support for Python USB devices.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

PythonLinks commented:
Thank you for working on this. 

I would like to use this for running MicroPython on the Pico-Ice, a board with an RP2040, and a Lattice ICE40UP5K FPGA.   The Pico-Ice has an SDK, which uses TinyUSB.  So this would allow both the Pico-Ice SDK and MicroPython to coexist. 

From my point of view, it is this huge mass of software, TinyUSB, MicroPython, and the Pico-ICE SDK which is very scary to get involved with.   Maybe I should just stay focused on my FPGA work.  On the other hand it would allow me to offer my FPGA signal coprocessor to the MicroPython community.   So eventually I need to get this to work.  And we do have a class project to do something with C and an FPGA UART, so maybe I could find a doable project in this large space.   Much more interesting than what the teacher proposed using the FTDI chip. 

My apologies if I am a newbie in this area.  Any guidance would be appreciated.   In the meantime I am going to read more.   Maybe starting with the TinyUSB documentation, the micropython USB documentation, and the Pico-ICE SDK source code, (since I have already read their excellent documentation).

thewh1teagle commented:
What's the status of this? Can I use HID with MicroPython on Raspsberry PI Pico W?

projectgus commented:
@PythonLinks Your project sounds exciting! I'm not sure exactly how this PR fits in, but if you have a goal that requires a custom USB interface that then goes via the RP2040 to the FPGA (and doesn't need maximum throughput) then this can probably help.

> What's the status of this? Can I use HID with MicroPython on Raspsberry PI Pico W?

@thewh1teagle You can, if you build MicroPython from the branch in this PR and then install some of the packages from the linked micropython-lib PR. There is a HID base class and example included in the micropython-lib PR.

patfrench commented:
Hello and happy new year

I'm working on an online micropython editor project. I'm facing several problems.

I'm using an RPi PICO program with the firmware of this fork 
[https://github.com/patfrench/micropython-rp2-webusb](https://github.com/patfrench/micropython-rp2-webusb)

_see the PR_ [https://github.com/micropython/micropython/pull/11301#top](https://github.com/micropython/micropython/pull/11301#top)

The board is recognized as a virtual port com but also as a Universal serial bus peripheral (with the WINUSB driver).

![cinq](https://github.com/micropython/micropython/assets/44723412/3bc72120-eaec-4b95-b513-40a7e282f697)

![un](https://github.com/micropython/micropython/assets/44723412/833d24ec-83f9-461a-9512-fa2c221a8d3a)

![deux](https://github.com/micropython/micropython/assets/44723412/4f9a3c3f-aabf-41b1-958a-c17d53c91d9d)

![trois](https://github.com/micropython/micropython/assets/44723412/105d9f07-417e-4c2e-bc24-da1b26416c87)

![six](https://github.com/micropython/micropython/assets/44723412/8fb2abc3-f529-4564-96d0-fa648656ea9b)
I can also see the REPL on the VCP COM6

![quatre](https://github.com/micropython/micropython/assets/44723412/eafb1fe5-4d8a-46d1-b61c-7db79e0efdcf)

In the TERMINAL tab there's a pyscript console because I'd like to use the power of pyscript

I've rebuilt this fork in version 1.22 preview (see the REPL micropython) but now I can't make Micropython evolve because something broke in the compilation and I don't know enough about mechanics so I'm stuck.

So I wanted to try the usbd project from @projectgus 

I ran cdc_repl_example. py, my PC created a second COM PORT (COM22) but I can't connect to it. I get this error message

![sept](https://github.com/micropython/micropython/assets/44723412/27d932e0-30fc-4e27-9d22-89875d2e0af0)

1. Where do I start to make my card WebUSB compatible?

2. cdc_repl_example.py creates a second COM port how can I change the descriptor to that of a WebUSB card?

I prefer @projetgus' method (writing drivers in python) rather than the firmware I use today.
I'd like to follow Micropython's evolution and not have my own fork.

Thank you very much

projectgus commented:
Rebased this PR just now, no major changes.

@patfrench Thanks for testing this PR out! Due to some recent changes in the rest of MP, `cdc_repl_example.py` currently only works if the main REPL is switched to the plain UART peripheral (`#define MICROPY_HW_USB_CDC (0)` and `#define MICROPY_HW_ENABLE_UART_REPL (1)`. I'm still debugging this, hopefully it will be possible to have them work concurrently. This is only because they are both CDCs with the REPL.

There also seems to be a more general bug with the CDC class on Windows, it's not receiving bytes as expected. Debugging this, also.

> cdc_repl_example.py creates a second COM port how can I change the descriptor to that of a WebUSB card?

We don't have an example for this, but it should be possible to implement the descriptors for WebUSB. Your goal is to run a MicroPython REPL over the second WebUSB interface (with WinUSB driver) as well, yes?

patfrench commented:
Hi,

> @patfrench Thanks for testing this PR out! Due to some recent changes in the rest of MP, cdc_repl_example.py currently only works if the main REPL is switched to the plain UART peripheral (#define MICROPY_HW_USB_CDC (0) and #define MICROPY_HW_ENABLE_UART_REPL (1). I'm still debugging this, hopefully it will be possible to have them work concurrently. This is only because they are both CDCs with the REPL.

All right, I'll try

> We don't have an example for this, but it should be possible to implement the descriptors for WebUSB. Your goal is to run a MicroPython REPL over the second WebUSB interface (with WinUSB driver) as well, yes?

today the same REPL is on 

- USB CDC with the driver usbser.sy. (claimed by the windows)
- USB CDC with the driver winusb (claimed by the chrome or edge)

With the USB of windows i use (putty, mpremote ...)
With the USB of web browser i use my editor ( API webusb javascript)

If tomorrow I had a REPL on a hardware UART and the same REPL on a winusb USB CDC, that would be perfect.

Thank you

patfrench commented:
Hi,
I've started some experiments (at my level)

- I ran cdc_repl_example.py

- my PC created 2 virtual com ports 

- I forced the installation of the WINUSB driver with ZADIG (on the new virtual com port)

- I compared the 2 PICO boards with chrome://usb-internals/

I need help because I can't find in your code the data I see in the images below.
Can you explain the mechanics to me? I'd like to understand to help (if I can)

Thank you,

Configuration with cdc_repl_example. py
![PICO_repl_cdc_example](https://github.com/micropython/micropython/assets/44723412/eed306f1-fecd-4d82-8735-9862f33635f3)

Configuration with my board 
![PICO_webUsb](https://github.com/micropython/micropython/assets/44723412/d98dd0a9-b767-41ee-aca5-0ccd4775396b)

Descriptor with cdc_repl_example.py
![PICO_repl_cdc_example_descriptor](https://github.com/micropython/micropython/assets/44723412/599f6be0-9c18-4837-beec-62d343d7113d)

Descriptor of my board
![PICO_webUsb_descriptor](https://github.com/micropython/micropython/assets/44723412/aabc3af9-2267-47a0-a81d-afc318249ef6)

projectgus commented:
@patfrench I've just pushed a fix to the micropython-lib branch for an issue that I was seeing with USB-CDC on Windows, it may help you as well.

The current Python USB API is designed around the idea of adding interfaces to the USB device. As both WebUSB and the 
Microsoft OS compatibility descriptors don't associate directly with an interface (to my knowledge), but instead responds to BOS descriptor requests, there will be some change needed here to expose the TinyUSB callback `tud_descriptor_bos_cb` to a Python function.

projectgus commented:
There was a commit in this PR to defer initialisation of TinyUSB until after boot.py ran. This has been spun out to its own PR, #13571.

projectgus commented:
@ArnaudV27 The micropython-rp2-webusb fork you've linked is a very different approach to this PR (patfrench is working to migrate from using the fork to using this PR, this isn't yet possible.) Suggest talking to the author of the fork about this problem, it's not related to anything here.

bobtedbob commented:
Is there any example python code for this PR?  Ideally I'd like to be able to add my own descriptor and send data. but I don't know if this is possible yet

projectgus commented:
@bobtedbob There's no example directly targeting this low-level API, but there is in the companion micropython-lib PR: https://github.com/micropython/micropython-lib/pull/558

projectgus commented:
Have just pushed a major refactor based on some suggestions from @dpgeorge. Any existing code will need to be updated for this new version:

* The `machine.USBD.rst` file in this PR is up to date with docs for the low-level API.
* The [micropython-lib PR](https://github.com/micropython/micropython-lib/pull/558) is updated and uses this new API.

Any questions, concerns, or something is not working then please comment!

projectgus commented:
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.

projectgus committed:
rp2: Enable support for Python USB devices.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
mimxrt: Define the MICROPY_HW_ENABLE_USBDEV macro.

Previously USB was always enabled, but this created some conflicts when
adding guards to other files on other ports.

Note the configuration with USB disabled hasn't been tested and probably
won't build or run without further work.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
Pushed another big update with some breaking API changes that came up in review. Check the included docs for details.

Have also updated the size numbers in the first post.

This update also squashes everything down to be merge-ready.

dpgeorge commented on shared/tinyusb/mp_usbd.h:
```diff
@@ -27,25 +27,117 @@
+    mp_obj_t control_xfer_cb;
+    mp_obj_t xfer_cb;
+
+    mp_usbd_builtin_driver_t builtin_driver;
```
Could this just be `mp_obj_t`, and it just stores the pointer to the current built-in driver object?

Then you don't need an enum, and the logic to get/set the attribute for this is very simple.

dpgeorge replied:
I think it's simpler to have the extern consts than the enum and the enum<->const mapping.

projectgus replied:
Done, PTAL!

dpgeorge commented on extmod/machine_usb_device.c:
```diff
@@ -0,0 +1,351 @@
+        // Load attribute.
+        if (attr == MP_QSTR_builtin_driver) {
+            // Currently simple as only two built-in drivers supported
+            switch (self->builtin_driver) {
```
If this was not an enum, and an `mp_obj_t`, the logic here could be simply `return self->builtin_driver`.

dpgeorge commented on ports/rp2/mpconfigport.h:
```diff
@@ -52,6 +52,11 @@
+
+#ifndef MICROPY_HW_ENABLE_USB_RUNTIME_DEVICE
+// Enable to support machine.USBDevice
+#define MICROPY_HW_ENABLE_USB_RUNTIME_DEVICE    (0)
```
IMO we should enable this by default.  If the feature is not used then users will not notice any difference.

dpgeorge commented on ports/samd/mpconfigport.h:
```diff
@@ -62,6 +62,11 @@
+
+#ifndef MICROPY_HW_ENABLE_USB_RUNTIME_DEVICE
+// Enable to use the 'usbd' module
+#define MICROPY_HW_ENABLE_USB_RUNTIME_DEVICE    (0)
```
IMO should enable by default.

dpgeorge commented on extmod/machine_usb_device.c:
```diff
@@ -0,0 +1,343 @@
+            o->xfer_data[i][0] = mp_const_none;
+            o->xfer_data[i][1] = mp_const_none;
+        }
+        o->builtin_driver = MP_OBJ_TO_PTR(&mp_type_usb_device_builtin_none);
```
This will need to be `MP_OBJ_FROM_PTR`, because you want to create an object.

projectgus replied:
Oops, I think that'll be the case in a few places...

dpgeorge commented on extmod/machine_usb_device.c:
```diff
@@ -0,0 +1,343 @@
+            if (
+                dest[1] == MP_OBJ_TO_PTR(&mp_type_usb_device_builtin_none)
+                #if HAS_BUILTIN_DRIVERS
+                || dest[1] == MP_OBJ_TO_PTR(&mp_type_usb_device_builtin_default)
```
You don't really need this check for a valid builtin.  The code in this PR will still work perfectly fine if you store any arbitrary object in `self->builtin_driver`.

I can see why you'd want some error checking, but is it needed?

projectgus replied:
The user won't get the behaviour they expect if they assign a different object here, the other pieces of code only check for "none" or "default" for built-in drivers. It doesn't look at any of the fields on the object from the C side.

That's something that maybe could be changed in the future by encoding more information about the drivers inside the object itself, so an arbitrary object would work as expected, but currently the logic that reads this field is comparing the address.

Or do you mean it's better to accept anything here and perhaps behave unexpectedly, rather than error out?

dpgeorge replied:
> Or do you mean it's better to accept anything here and perhaps behave unexpectedly, rather than error out?

Yes.  I'm just thinking from a small/minimal code size point of view.  This driver is low-level so doesn't need to be 100% user friendly.

projectgus replied:
Got it, have changed

dpgeorge commented:
Merged!!

Thanks @projectgus for all your efforts on this.  It's a great feature, will be widely used.

Luckinber commented:
Congrats @projectgus!

Super exciting to see this move forward, I'll see you in the lib pull request 😉

hoihu commented:
terrific feature! Congrats!
